chasm/lib/scheduler/scheduler.go

```diff
@@ -0,0 +1,275 @@
+		BackfillRequest: request,
+	}
+	s.addBackfiller(ctx, backfiller)
+	return backfiller
+}
+
+// NewImmediateBackfiller returns an intialized Backfiller component, which should
+// be parented under a Scheduler root node.
+func (s *Scheduler) NewImmediateBackfiller(
+	ctx chasm.MutableContext,
+	request *schedulepb.TriggerImmediatelyRequest,
+) *Backfiller {
+	backfiller := newBackfiller(ctx, s)
+	backfiller.BackfillerInternal.Request = &schedulespb.BackfillerInternal_TriggerRequest{
+		TriggerRequest: request,
+	}
+	s.addBackfiller(ctx, backfiller)
+	return backfiller
+}
+
+// addBackfiller adds the backfiller to the scheduler tree, and adds a task to
+// kick off backfill processing.
+func (s *Scheduler) addBackfiller(
+	ctx chasm.MutableContext,
+	backfiller *Backfiller,
+) {
+	s.Backfillers[backfiller.BackfillId] = chasm.NewComponentField(ctx, backfiller)
+	ctx.AddTask(backfiller, chasm.TaskAttributes{}, &schedulespb.BackfillerTask{})
+}
+
+// useScheduledAction returns true when the Scheduler should allow scheduled
+// actions to be taken.
+//
+// When decrement is true, the schedule's state's `RemainingActions` counter is
+// decremented when an action can be taken. When decrement is false, no state
+// is mutated.
+func (s *Scheduler) useScheduledAction(decrement bool) bool {
+	scheduleState := s.Schedule.State
+
+	// If paused, don't do anything.
+	if scheduleState.Paused {
+		return false
+	}
+
+	// If unlimited actions, allow.
+	if !scheduleState.LimitedActions {
+		return true
+	}
+
+	// Otherwise check and decrement limit.
+	if scheduleState.RemainingActions > 0 {
+		if decrement {
+			scheduleState.RemainingActions--
+
+			// The conflict token is updated because a client might be in the process of
+			// preparing an update request that increments their schedule's RemainingActions
+			// field.
+			s.updateConflictToken()
+		}
+		return true
+	}
+
+	// No actions left
+	return false
+}
+
+func (s *Scheduler) getCompiledSpec(specBuilder *scheduler.SpecBuilder) (*scheduler.CompiledSpec, error) {
+	s.validateCachedState()
+
+	// Cache compiled spec.
+	if s.compiledSpec == nil {
+		cspec, err := specBuilder.NewCompiledSpec(s.Schedule.Spec)
+		if err != nil {
+			return nil, err
+		}
+		s.compiledSpec = cspec
+	}
+
+	return s.compiledSpec, nil
+}
+
+func (s Scheduler) jitterSeed() string {
+	return fmt.Sprintf("%s-%s", s.NamespaceId, s.ScheduleId)
+}
+
+func (s Scheduler) identity() string {
+	return fmt.Sprintf("temporal-scheduler-%s-%s", s.Namespace, s.ScheduleId)
+}
+
+func (s Scheduler) overlapPolicy() enumspb.ScheduleOverlapPolicy {
+	policy := s.Schedule.Policies.OverlapPolicy
+	if policy == enumspb.SCHEDULE_OVERLAP_POLICY_UNSPECIFIED {
+		policy = enumspb.SCHEDULE_OVERLAP_POLICY_SKIP
+	}
+	return policy
+}
+
+func (s Scheduler) resolveOverlapPolicy(overlapPolicy enumspb.ScheduleOverlapPolicy) enumspb.ScheduleOverlapPolicy {
+	if overlapPolicy == enumspb.SCHEDULE_OVERLAP_POLICY_UNSPECIFIED {
```

Comment:
Why don't these methods take a pointer receiver? I'm surprised we don't have lint errors for this.